api/s2n.h

```diff
@@ -687,6 +687,266 @@ extern int s2n_cert_get_utf8_string_from_extension_data_length(const uint8_t *ex
 S2N_API 
 extern int s2n_cert_get_utf8_string_from_extension_data(const uint8_t *extension_data, uint32_t extension_len, uint8_t *out_data, uint32_t *out_len);
 
+/* Pre-shared key (PSK) Hash Algorithm - RFC 8446 Section-2.2 */
+typedef enum {
+    S2N_PSK_HMAC_SHA256,
+    S2N_PSK_HMAC_SHA384,
+} s2n_psk_hmac;
+
+struct s2n_psk;
+
+/**
+ * Creates a new s2n external Pre-shared key (PSK) object with `S2N_PSK_HMAC_SHA256` as the default 
+ * PSK hash algorithm. An external PSK is a key established outside of TLS using a secure mutually agreed upon mechanism.
+ * 
+ * Use `s2n_psk_free` to free the memory allocated to the s2n external PSK object created by this API. 
+ *
+ * @return struct s2n_psk* Returns a pointer to the newly created external PSK object.
+ */
+S2N_API
+struct s2n_psk* s2n_external_psk_new();
+
+/**
+ * Frees the memory associated with the external PSK object.
+ *
+ * @param psk Pointer to the PSK object to be freed.
+ */
+S2N_API 
+int s2n_psk_free(struct s2n_psk **psk);
+
+/**
+ * Sets the identity for a given external PSK object.
+ * The identity is a unique identifier for the pre-shared secret.
+ * It is a non-secret value represented by raw bytes.
+ *
+ * # Safety 
+ *
+ * The identity is transmitted over the network unencrypted and is a non-secret value.
+ * Do not include confidential information in the identity.
+ * 
+ * Note that the identity is copied into s2n-tls memory and the user is responsible for freeing the memory 
+ * associated with the identity input. 
+ *
+ * @param psk A pointer to a PSK object to be updated with the identity.
+ * @param identity The identity in raw bytes format to be copied.
+ * @param identity_size The length of the PSK identity being set.
+ */
+S2N_API 
+int s2n_psk_set_identity(struct s2n_psk *psk, const uint8_t *identity, uint16_t identity_size);
+
+/**
+ * Sets the out-of-band/externally provisioned secret for a given external PSK object.
+ *
+ * # Safety
+ *
+ * Note that the secret is copied into s2n-tls memory and the user is responsible for freeing 
+ * the memory associated with the `secret` input. 
+ *
+ * Deriving a shared secret from a password or other low-entropy source
+ * is not secure and is subject to dictionary attacks.
+ * See https://tools.ietf.org/rfc/rfc8446#section-2.2 for more information.
+ *
+ * @param psk A pointer to a PSK object to be updated with the secret.
+ * @param secret The secret in raw bytes format to be copied.
+ * @param secret_size The length of the pre-shared secret being set.
+ */
+S2N_API 
+int s2n_psk_set_secret(struct s2n_psk *psk, const uint8_t *secret, uint16_t secret_size);
+
+/**
+ * Sets the hash algorithm for a given external PSK object. The supported PSK Hash 
+ * algorithms are as listed in the enum `s2n_psk_hmac` above.
+ * 
+ * @param psk A pointer to the external PSK object to be updated with the PSK Hash algorithm.
+ * @param hmac The PSK Hash algorithm being set.  
+ */
```

Comment:
We should be careful switching between HMAC / Hash here. If you want to make a distinction between the two instead of just using HMAC, then make it very explicitly.